spv/sync.go

```diff
@@ -425,10 +425,28 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 		return nil
 	})
 
-	// Run wallet background goroutines (currently, this just runs
-	// mixclient).
+	// Ensure wallet.Run cleanly finishes/is canceled first when outer
+	// context is canceled.
+	walletCtx, walletCtxCancel := context.WithCancel(context.Background())
+	go func() {
+		select {
+		case <-ctx.Done():
+		case <-gctx.Done():
+		}
+		walletCtxCancel()
+	}()
 	g.Go(func() error {
-		return s.wallet.Run(ctx)
+		// Run wallet background goroutines (currently, this just runs
+		// mixclient).
+		err := s.wallet.Run(walletCtx)
+		if err != nil {
+			return err
+		}
+
+		// If gctx has not yet been canceled, do so here now.
+		// walletCtx is canceled after either ctx or gctx is canceled.
+		<-walletCtx.Done()
```

Comment:
This doesn't look right to me yet.

In the case an error is returned from `Run` unrelated to the context being canceled, that error would be returned which would cause the group context to be canceled so everything would be fine.

However, when `wallet.Run` returns `nil`, `walletCtx` will never be canceled afaict.  Thus, it would just hang here.

Reply:
It will be cancelled by the goroutine started above this.

Reply:
https://go.dev/play/p/5VPuDP_Bbtw

Reply:
That is assuming there is an error in the goroutines in the group though.  There won't necessarily be one.

Reply:
https://go.dev/play/p/sYabLXOyzoN

Reply:
but they all return errors....

have you ran this and seen it hang?

Reply:
I haven't seen any issues, but I'm just considering the edge cases.  So long as there is a never a case where the grouped goroutines all exit without an error, then the code is fine.

My worry is just basically what I described.  Namely, `wallet.Run` returns `nil` and then it just sits and waits, effectively until either there is an error in the group or the outer context is canceled.  That part is correct.

However, it implies all of the goroutines in the group will be sitting there doing nothing _after_ `wallet.Run` returns until eventually the outer context is canceled even though they really should be exiting after `wallet.Run` returns unconditionally so that `syncer.Run` returns without error versus sitting around in a defunct state.

Reply:
Ok, after looking through more of the surrounding code, I guess it's just the poor naming of things that misled me and those other goroutines are supposed to remain active until the outer context is canceled regardless of whether or not the "wallet" is running.  It'll still be processing p2p messages, fetching filters, etc.

I was completely misled by the naming to believe it was the entire **wallet** as opposed to some optional additional optional background tasks.